app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,148 @@
+
+    init {
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private var metricDataContainer = AppStartupMetricDataContainer()
+
+    private var isMetricRecordedSinceAppWasForegrounded = false
+    private var wasAppCreateCalledBeforeActivityCreate = false
+
+    fun onFenixApplicationOnCreate() {
+        wasAppCreateCalledBeforeActivityCreate = true
+    }
+
+    fun onHomeActivityOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        metricDataContainer.hasSavedInstance = hasSavedInstanceState
+        metricDataContainer.type = getAppStartupType()
+        metricDataContainer.source = getStartupSourceFromIntent(safeIntent, false)
+
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    fun onExternalAppBrowserOnCreate(safeIntent: SafeIntent, hasSavedInstanceState: Boolean) {
+        metricDataContainer.hasSavedInstance = hasSavedInstanceState
+        metricDataContainer.type = getAppStartupType()
+        metricDataContainer.source = getStartupSourceFromIntent(safeIntent, true)
+
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    /**
+     * we are not setting [Source] in this method since source is derived from a [SafeIntent].
+     * therefore source either gets set in oncreate() or onNewIntent().
+     */
+    fun onHomeActivityOnRestart() {
+        metricDataContainer.type = Type.HOT
+        metricDataContainer.hasSavedInstance = false
+    }
+
+    /**
+     * we are only setting [Source] in this method since source is derived from a [SafeIntent].
+     * other metric fields are set in onCreate() and onRestart()
+     */
+    fun onHomeActivityOnNewIntent(safeIntent: SafeIntent) {
+        metricDataContainer.source = getStartupSourceFromIntent(safeIntent, false)
+    }
+
+    private fun getAppStartupType(): Type {
+        return if (wasAppCreateCalledBeforeActivityCreate) COLD else WARM
+    }
+
+    private fun getStartupSourceFromIntent(
+        intent: SafeIntent,
+        isExternalAppBrowserActivity: Boolean
+    ): Source {
+        return when {
+            isExternalAppBrowserActivity -> CUSTOM_TAB
+            intent.isLauncherIntent -> APP_ICON
+            intent.action == Intent.ACTION_VIEW -> LINK
+            else -> ERROR
+        }
+    }
+
+    /**
+     * The reason we record metric on resume is because we need to wait for onNewIntent(), and
+     * we are not guaranteed that onNewIntent() will be called before or after onStart() / onRestart().
+     * However we are guaranteed onResume() will be called after onNewIntent() and  onStart(). Source:
+     * https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)
+     *
+     */
+    fun onHomeActivityOnResume() {
+        recordMetric()
+    }
+
+    private fun recordMetric() {
+        if (!isMetricRecordedSinceAppWasForegrounded) {
+            metrics.track(metricDataContainer.getAppStartupSourceAndType())
```

Comment:
This implementation has similar problems to the previous implementation:
- it's not explicit how the data from `onRestart` and `onNewIntent` are joined
- we don't handle, and document, the app_switcher case
- we don't handle, and document, the error case too explicitly

For this method, I'm essentially thinking:
```kotlin
if (!isMetricRecordedSinceAppWasForegrounded) {
    if (onResumeData != null && onNewIntentData != null) {
        // create immutable & track
    } else if ( // handle other cases such as app switcher
```

It explicitly documents our assumption (that `onNewIntent` and `onResume` should be called together) and explicitly documents each of the possible cases in a central location.

Reply:
fun getAppStartupSourceAndType(): AppAllStartup {
            setNullFieldsToError()
            return AppAllStartup(source!!, type!!, hasSavedInstance!!)
        }

is essentially where I'm doing that. I think we decided to send `ERROR` if a field is not set? 

`setNullFieldsToError()` does seem to be hidden away, and maybe we need to call this inside `recordMetric()`
so something like 

```
if (!isMetricRecordedSinceAppWasForegrounded) {
            setNullFieldsToError()
           track(event....)
}

```
Doing callback specific `if`  statement seems a bit weird since the same data is recorded in multiple states (eg `onCreate` records all the data)